### DISCUSSION TITLE
Continuation: Image resizing with LCD sub-pixel awareness (aka ClearType)

### DISCUSSION THREAD

**Discussion by nick87720z** [ImageMagick/ImageMagick#5859] [Digital Image Processing]
This is for topic from legacy forum: https://imagemagick.org/discourse-server/viewtopic.php?t=19120

*I have old account with password last change in 2010, but probably lost access - recovery doesn't work for both of my emails, yet there's not "register" button (so I could check if my account there still exists).*

First I enhanced original processing using imagemagick. Considering same freetype, it has 2 useful and simple to understand filters: default and light.
- Light is as simple as needs to be - 3x box, which is all you need to make subpixels (really - HiRes pixels behind subs) fully color capable.
- Default is almost same. Although it's spreads some portion from edge elements further (to 5 subs width), it's too far from bilinear interpolation (1,2,3,2,1 from original 3,3,3) and only addresses imperfections in real-world displays. Nothing supernatural.

However, try such experiment: somehow make screenshot in 3x resolution (I propose to use `xrandr --scale` feature) with font greyscale antialiasing (no-subpixel fontconfig conf), then try to manually shift channels and apply that simple 3x box (just as FT lcdlight filter). You would expect fonts to be as clear as in this one, but instead - there would look blured even more than with FT lcddefault.

Reason is simple: 3x box loses some contrast along with level of detail.  

**Solution.**

If you tried `fft->filter->ifft`, this would give best `contrast / color fringes` ratio. However, fft/ifft are too expensive for real use. I only used that to prepare wallpapers. Still there's solution. It could be possible to make a kernel from fft/ifft filtering transform against single point. 
**Update:** It seems, kernelizing single sample, filtered with fft, represents nothing, but lanczos filter, which is already here (lol). The only advantage using fft/ifft is that it can remove higher frequencies without touching lower frequencies, thus doing enough filtering to later downscale with `-filter=point` (or `-sample`).

And then - lanczos appeared to be not enough good, because all I needed is single support shadows, adding right how much contrast was lost with 3x box, without extra ringing. Also, if trying to do it in linear space (with gamma correction, to avoid color fringes), darker wave parts become more visible (easy to notice if trying fft..ifft chain while in gamma-corrected mode).

I tried to conterweight blur with slight sharpening. Very small, using kernel `[-1,27,-1]` at the pixel scale. For subpixel scale it is `[-1,-1,-1, 27,27,27, -1,-1,-1]`, however I figured out, this is overkill - pointless waste of power. Although this filtering makes some disbalance (center subpixel in 3/3px wide lines overweight, producing color tint)

There's script with this enhancement (it's ready to use, taking exactly 4 args - target width and height, input and output files).
```sh
#!/bin/sh
#
# Downsampling with subpixel filtering for horizontal RGB

if [ $# -ne 4 ]; then
  echo "Usage: $0 <dest_width> <dest_height> <input_file> <output_file>"
  exit -1
fi
input="$3"
output="$4"
tw=$(( $1 ))
th=$(( $2 ))

# Output size check & guess

eval "$( identify "${input}" | awk -vtw="$tw" -vth="$h" '{
  FS="x"
  $0=$3
  iw = $1, ih = $2
  if ( tw == 0)
  {
    if (th == 0 )
      tw = int( iw / 3 ), th = int( ih / 3 );
    else
      tw = int( iw * th / ih );
  }
  else if (th == 0)
      th = ih * tw / iw;
  print "iw="$1"; ih="$2;
}')"

if [ $tw -eq 0 ] || [ $th -eq 0 ]; then
  echo "Image has zero total size. Aborting."
  exit -1
fi

echo "Target size: $tw x $th"

# Stages, in order
# - vertical HQ resize
# - horizontal 3x box (to make each subpixel full color-capable)
# - RGB channels shift (combining surrounding subpixels in one ready to use pixel)
# - sharpening: just enough to restore original contrast (persuing FFT/IFFT ideal)

magick -verbose "${input}" -depth 32 -gamma 0.5 \
-filter catrom -resize $(( tw * 3 ))x$((th))\! \
-convolve "3,1: 1,1,1" \
-channel red  -convolve "3,1:1,0,0" \
-channel blue -convolve "3,1:0,0,1" +channel \
-filter Point -resize $((tw))x$((th))\! \
-gamma 2.0 \
-convolve "3,1: -1,27,-1" \
-depth 8 "${output}"
```
---
Besides, that thread ended with expression of hope, that it could be accelerated in e.g. OpenGL. Just now picom compositing manager (xcompmrg / compton continuation) has stable support for custom post-processing shaders. Unfortunally it's currently limited only for windows, screen post-processing is not (yet) supported. But at least it works for windows.

So, I made lcd filtering shader, with only difference, that fragment shaders are not allowed to resize, only filter (however, this is not big problem, since shaders can evaluate coordinates and thus, bypass their processing based in coord values).

Link to shader: https://github.com/yshui/picom/wiki/Shader-Shop#fullscreen-subpixel-antialiasing

**Edit:** I edited some parts above, because only just (on these days) discovered, why gamma correction is important for subpixel filtering. Also I originally did a mistake in fft..ifft way, setting cutoff frequency to `image_width / 3`, while it had to be `image_width / 2 / 3` (which is why I got too good contrast, it just clearly had to had dozens of color fringes with certain input).

**Comment by urban-warrior** [comment]
It certainly would be simple enough to convert your script to a single command-line option.  Any suggestions for the option name, e.g., `-subpixel-rendering`?

**Reply by nick87720z** [reply]
Is it for magick command? Could be also --subpixel-filter, to just filter (without downscaling) or --subpixel-scale (benefiting from skiping unnecessary pixels).

Also possible --filter=subpixel, but if used for --resize, it would still need something for resampling.

**Comment by nick87720z** [comment]
I'm still in uncertainty, what's best way to make a sharpening. If it's done in linear space (between gamma correction clamps), then dark areas get visible contrast support, while for light areas they are fine *(also happening if using lanczos-type downscale or just fft..ifft in linear, which seem to be same in result)*. This also may be reason, why light on dark lines become thicker than dark on light.

However, if contrast is added after exit from linear color space, this causes similar balance break as would subpixel filtering do if done not in linear too: lines 3/3px wide (in destination units) got central subpixel overweight, thus getting proper tint *(green for unshifted and blue for 1/3px shift, for HRGB layout)*.
